assets/js/main.js

```diff
@@ -0,0 +1,167 @@
+(function() {
+  // ================== globals ==================
+  var COLORS = ['red', 'green', 'blue', 'yellow'];
+  var START_DIFFICULTY = 3;
+  var COLOR_LIGHT_TIME = 300;
+  var STARTING_TIME_PER_COLOR = 1000;
+
+  var sounds = {};
+
+  var $colors = {};
+  var $simon = null;
+  var $score = null;
+
+  var game = null;
+
+  // ================== utility ==================
+  // Returns a random integer between min (included) and max (excluded)
+  // Using Math.round() will give you a non-uniform distribution!
+  var getRandomInt = function(min, max) {
+    return Math.floor(Math.random() * (max - min)) + min;
+  }
+
+  // returns a random element from an array
+  var getRandomFromArray = function(arr) {
+    var randIndex = getRandomInt(0, arr.length);
+    return arr[randIndex];
+  };
+
+  // ================== game functions ==================
+
+  // setup the global jquery variables
+  var setup = function(){
+    for(var i = 0; i < COLORS.length; i = i + 1){
+      var color = COLORS[i];
+      setupColor(color);
+    }
+    $simon = $('.simon');
+    $score = $('.score');
+  };
+
+  // setup one colors jquery variable and click listener
+  var setupColor = function(color){
+    var $color = $('.' + color);
+    $colors[color] = $color;
+    sounds[color] = new Audio('assets/sound/' + color + '.mp3');
+    // currying maybe too complex, alternately we can use the target
+    $color.click(function() { onColorClick(color); });
+  };
+
+  // restart the game
+  var restart = function(colors) {
+    game = {
+      pattern: null,
+      score: 0,
+      difficulty: START_DIFFICULTY,
+      patternIndex: 0,
+      clickable: false,
+      timePerColor: STARTING_TIME_PER_COLOR
+    };
+    $score.html(game.score);
+    nextPattern();
+  };
+
+  // generate he next pattern and play it
+  var nextPattern = function(){
+    // this is to prevent the user from clicking on stuff until it's time
+    game.clickable = false;
+    $simon.toggleClass('clickable', false);
```

Comment:
I considered creating a bunch of functions that are game state changes tied to corresponding view changes but thought maybe that was too complex of an idea?